tensorflow/compiler/mlir/tensor/transforms/legalize_utils.cc

```diff
@@ -0,0 +1,249 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <algorithm>
+#include <iostream>
+#include <numeric>
+
+#include "mlir/Dialect/Arith/IR/Arith.h"  // from @llvm-project
+#include "mlir/Dialect/Linalg/IR/Linalg.h"  // from @llvm-project
+#include "mlir/Dialect/Math/IR/Math.h"  // from @llvm-project
+#include "mlir/Dialect/SCF/IR/SCF.h"  // from @llvm-project
+#include "mlir/Dialect/Tensor/IR/Tensor.h"  // from @llvm-project
+#include "tensorflow/compiler/mlir/tensor/transforms/legalize_utils.h"
+
+namespace mlir {
+namespace tensor {
+
+namespace {
+
+// Return the total size of the given tensor if it is ranked and all of its
+// dimensions have static size. Otherwise return nullopt. The given value is
+// expected to be a tensor.
+std::optional<int64_t> getTensorStaticSize(Value tensor) {
```

Comment:
Isn't this ShapedType's getNumElements with a return of null if not statically shaped?

Reply:
It is. Looking at the implementation of `ShapedType::getNumElements()` though, I see that it asserts if `num` becomes negative during the computation, which would happen if a dynamic dimension is present. I'm afraid I can't rely on that call then.